src/ESPeasyGPIO.ino

```diff
@@ -0,0 +1,393 @@
+  return digitalRead(pinNumber) == HIGH;
+}
+
+//********************************************************************************
+// MCP23017 read
+//********************************************************************************
+int8_t GPIO_MCP_Read(byte Par1)
+{
+  int8_t state = -1;
+  byte unit = (Par1 - 1) / 16;
+  byte port = Par1 - (unit * 16);
+  uint8_t address = 0x20 + unit;
+  byte IOBankValueReg = 0x12;
+  if (port > 8)
+  {
+    port = port - 8;
+    IOBankValueReg++;
+  }
+  // get the current pin status
+  Wire.beginTransmission(address);
+  Wire.write(IOBankValueReg); // IO data register
+  Wire.endTransmission();
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    state = ((Wire.read() & _BV(port - 1)) >> (port - 1));
+  }
+  return state;
+}
+
+
+//********************************************************************************
+// MCP23017 write
+//********************************************************************************
+boolean GPIO_MCP_Write(byte Par1, byte Par2)
+{
+  boolean success = false;
+  byte portvalue = 0;
+  byte unit = (Par1 - 1) / 16;
+  byte port = Par1 - (unit * 16);
+  uint8_t address = 0x20 + unit;
+  byte IOBankConfigReg = 0;
+  byte IOBankValueReg = 0x12;
+  if (port > 8)
+  {
+    port = port - 8;
+    IOBankConfigReg++;
+    IOBankValueReg++;
+  }
+  // turn this port into output, first read current config
+  Wire.beginTransmission(address);
+  Wire.write(IOBankConfigReg); // IO config register
+  Wire.endTransmission();
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    portvalue = Wire.read();
+    portvalue &= ~(1 << (port - 1)); // change pin from (default) input to output
+
+    // write new IO config
+    Wire.beginTransmission(address);
+    Wire.write(IOBankConfigReg); // IO config register
+    Wire.write(portvalue);
+    Wire.endTransmission();
+  }
+  // get the current pin status
+  Wire.beginTransmission(address);
+  Wire.write(IOBankValueReg); // IO data register
+  Wire.endTransmission();
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    portvalue = Wire.read();
+    if (Par2 == 1)
+      portvalue |= (1 << (port - 1));
+    else
+      portvalue &= ~(1 << (port - 1));
+
+    // write back new data
+    Wire.beginTransmission(address);
+    Wire.write(IOBankValueReg);
+    Wire.write(portvalue);
+    Wire.endTransmission();
+    success = true;
+  }
+  return(success);
+}
+
+
+//********************************************************************************
+// MCP23017 config
+//********************************************************************************
+void GPIO_MCP_Config(byte Par1, byte Par2)
+{
+  // boolean success = false;
+  byte portvalue = 0;
+  byte unit = (Par1 - 1) / 16;
+  byte port = Par1 - (unit * 16);
+  uint8_t address = 0x20 + unit;
+  byte IOBankConfigReg = 0xC;
+  if (port > 8)
+  {
+    port = port - 8;
+    IOBankConfigReg++;
+  }
```

Comment:
Is this just a single step, or should this be a while instead of an if?

Reply:
It's not my code. It's the original code.
I just moved it outside the plugin.